shader/antialiasing/cmaa2/cmaa2_common.glsl

```diff
@@ -0,0 +1,176 @@
+#endif
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// VARIOUS QUALITY SETTINGS
+//
+// Longest line search distance; must be even number; for high perf low quality start from ~32 - the bigger the number, 
+// the nicer the gradients but more costly. Max supported is 128!
+const uint c_maxLineLength = 86;
+// 
+
+layout(binding = 0) uniform sampler2D g_inoutColorReadonly;
+
+#if CMAA2_EDGE_UNORM 
+  layout(r8, binding = 2) uniform image2D g_workingEdges;
+#else
+  layout(r32ui, binding = 2) uniform uimage2D g_workingEdges;
+#endif
+
+layout(binding = 3) buffer UboWorkingShapeCandidates {
+  uint g_workingShapeCandidates[];
+  };
+
+layout(binding = 4) buffer UboWorkingDeferredBlendLocationList {
+  uint g_workingDeferredBlendLocationList[];
+  };
+
+layout(binding = 5) buffer UboWorkingDeferredBlendItemList {
+  uvec2 g_workingDeferredBlendItemList[];
+  };
+
+layout(r32ui, binding = 6) uniform uimage2D g_workingDeferredBlendItemListHeads;
+
+layout(binding = 7) buffer UboWorkingControlBuffer {
+  uint g_workingControlBuffer[];
+  };
+
+#define CmaaSaturate(x) clamp(x, 0.0, 1.0) 
+
+uvec4 UnpackEdges(uint value) {
+  uvec4 ret;
+  ret.x = uint((value & 0x01) != 0);
+  ret.y = uint((value & 0x02) != 0);
+  ret.z = uint((value & 0x04) != 0);
+  ret.w = uint((value & 0x08) != 0);
+  return ret;
+  }
+
+vec4 UnpackEdgesFlt(uint value) {
+  vec4 ret;
+  ret.x = float((value & 0x01) != 0);
+  ret.y = float((value & 0x02) != 0);
+  ret.z = float((value & 0x04) != 0);
+  ret.w = float((value & 0x08) != 0);
+  return ret;
+  }
+
+vec3 LoadSourceColor(ivec2 pixelPos, ivec2 offset, int sampleIndex) {
+  vec3 color = texelFetch(g_inoutColorReadonly, pixelPos + offset, 0).xyz;
+  return color;
+  }
+
+uint Pack_R11G11B10_FLOAT(vec3 rgb) {
+  rgb = min(rgb, uintBitsToFloat(0x477C0000));
+  uint r = ((floatBitsToUint(rgb.x) + 8) >> 4) & 0x000007FF;
+  uint g = ((floatBitsToUint(rgb.y) + 8) << 7) & 0x003FF800;
+  uint b = ((floatBitsToUint(rgb.z) + 16) << 17) & 0xFFC00000;
+  return r | g | b;
+  }
+
+vec3 Unpack_R11G11B10_FLOAT(uint rgb) {
+  float r = uintBitsToFloat((rgb << 4) & 0x7FF0);
+  float g = uintBitsToFloat((rgb >> 7) & 0x7FF0);
+  float b = uintBitsToFloat((rgb >> 17) & 0x7FE0);
+  return vec3(r, g, b);
+  }
+
+uint PackFloat32AsFloat16AndConvertToUint(float arg) {
+  return packHalf2x16(vec2(arg, 0.));
+  }
+
+float UnpackUintAsFloat16AndConvertToFloat32(uint arg) {
+  return unpackHalf2x16(arg).x;
+  }
+
+uint Pack_R11G11B10_E4_FLOAT(vec3 rgb) {
+  rgb = clamp(rgb, 0.0, uintBitsToFloat(0x3FFFFFFF));
+  uint r = ((PackFloat32AsFloat16AndConvertToUint(rgb.r) + 4) >> 3) & 0x000007FF;
+  uint g = ((PackFloat32AsFloat16AndConvertToUint(rgb.g) + 4) << 8) & 0x003FF800;
+  uint b = ((PackFloat32AsFloat16AndConvertToUint(rgb.b) + 8) << 18) & 0xFFC00000;
+  return r | g | b;
+  }
+
+vec3 Unpack_R11G11B10_E4_FLOAT(uint rgb) {
```

Comment:
any reason to support `R11G11B10_E4` in shader? Engine doesn't support it, and there is not plans to implements shared exponent formats.

Reply:
It's taken from https://github.com/Microsoft/DirectX-Graphics-Samples/blob/master/MiniEngine/Core/Shaders/PixelPacking_R11G11B10.hlsli

This format is used for packing LDR range more efficiently than standard R11G11B10 by stripping the exponent.